pkg/hub/lease/controller.go

```diff
@@ -0,0 +1,169 @@
+		// cluster is not accepted, skip it
+		acceptedCondition := helpers.FindSpokeClusterCondition(cluster.Status.Conditions, clusterv1.SpokeClusterConditionHubAccepted)
+		if !helpers.IsConditionTrue(acceptedCondition) {
+			continue
+		}
+
+		// check the cluster lease to determine the spoke cluster whether is available
+		availableCondition, err := c.checkLease(cluster)
+		if err != nil {
+			return err
+		}
+
+		availableConditionUpdateFn := helpers.UpdateSpokeClusterConditionFn(*availableCondition)
+		_, updated, err := helpers.UpdateSpokeClusterStatus(ctx, c.clusterClient, cluster.Name, availableConditionUpdateFn)
+		if err != nil {
+			return err
+		}
+		if updated {
+			syncCtx.Recorder().Eventf("SpokeClusterAvailableConditionUpdated", "update spoke cluster %q available condition to %q, due to %q",
+				cluster.Name, availableCondition.Status, availableCondition.Reason)
+		}
+	}
+	return nil
+}
+
+// checkLease checks a given cluster's lease and determine the cluster whether is available.
+// 1. if the cluster does not have a lease, update the cluster available condition to unknown
+// 2. if the cluster have a lease and does not have an available condition, update the cluster available condition to unknown
+// 3. if the cluster have a lease and have an available condition
+//   - if the lease is constantly updated, keep the cluster available condition
+//   - if the lease is not constantly updated in a duration, update the cluster available condition to unknown
+func (c *leaseController) checkLease(cluster *clusterv1.SpokeCluster) (*clusterv1.StatusCondition, error) {
+	clusterLeaseName := fmt.Sprintf("cluter-%s-lease", cluster.Name)
+	observedLease, err := c.leaseLister.Leases(cluster.Name).Get(clusterLeaseName)
+	if errors.IsNotFound(err) {
+		return &clusterv1.StatusCondition{
+			Type:    clusterv1.SpokeClusterConditionAvailable,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "SpokeClusterLeaseNeverUpdated",
+			Message: "Registration agent never updated its lease.",
+		}, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	observedCond := helpers.FindSpokeClusterCondition(cluster.Status.Conditions, clusterv1.SpokeClusterConditionAvailable)
+	if observedCond == nil {
+		return &clusterv1.StatusCondition{
+			Type:    clusterv1.SpokeClusterConditionAvailable,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "SpokeClusterAvailableStatusNeverUpdated",
+			Message: "Registration agent never posted its available status.",
+		}, nil
+	}
+
+	lastClusterLease := c.clusterLeaseMap.get(clusterLeaseName)
+	// update the lease if it is constantly updated
+	if lastClusterLease.lease == nil || lastClusterLease.lease.Spec.RenewTime.Before(observedLease.Spec.RenewTime) {
+		lastClusterLease.lease = observedLease
+		lastClusterLease.probeTimestamp = metav1.Now()
+		c.clusterLeaseMap.set(clusterLeaseName, lastClusterLease)
+	}
+
+	if metav1.Now().After(lastClusterLease.probeTimestamp.Add(c.leaseUpdateGracePeriod)) {
```

Comment:
I think leaseUpdateGracePeriod will relate to leaseDuration set on spokecluster?

Reply:
using 5 times lease duration as the grace period